hippod/cache_achievements.py

```diff
@@ -1,27 +1,96 @@
 import os
 import json
+import datetime
+
+import pprint
+
 
 
 class Cache(object):
     def __init__(self, app, frequency):
         self.app = app
         self.frequency = frequency
         self.data = self.init_data()
+        self.stored_data = self.read_old_cache()
+        dt = datetime.datetime.now()
+        dt = dt.replace(hour=0, minute=0, second=0, microsecond= 0)
+        self.date = dt.timestamp()
+        self.stored_data['passed'].insert(len(self.stored_data['passed']), [self.date, 0])
+        self.stored_data['failed'].insert(len(self.stored_data['failed']), [self.date, 0])
+        self.stored_data['nonapplicable'].insert(len(self.stored_data['nonapplicable']), [self.date, 0])
 
 
     def init_data(self):
         d = dict()
         d['achievement-results-by-category'] = dict()
+        d['achievement-results-sunburn-chart'] = dict()
+        d['achievement-results-sunburn-chart']['name'] = 'init'
+        d['achievement-results-sunburn-chart']['children'] = list()
+        d['achievement-results-sunburn-chart']['root'] = 'root'
         return d
 
 
+    def read_old_cache(self):
+        cache_db = os.path.join(self.app['DB_CACHE_PATH'])
+        achievements_cache_path = os.path.join(cache_db, 'cache-achievements.db')
+        if not os.path.isfile(achievements_cache_path):
+            with open(achievements_cache_path, 'w') as file:
+                data = dict()
+                data = json.dumps(data, sort_keys=True,indent=4, separators=(',', ': '))
+                file.write(data)
+        with open(achievements_cache_path, 'r') as file:
+            data = json.load(file)
+        if 'achievements-by-time' not in data:
+            self.data['achievements-by-time'] = dict()
+            self.data['achievements-by-time']['passed'] = list()
+            self.data['achievements-by-time']['failed'] = list()
+            self.data['achievements-by-time']['nonapplicable'] = list()
+        return self.data['achievements-by-time']
+
+
     def write_cache_file(self):
         cache_db = self.app['DB_CACHE_PATH']
         achievements_cache_path = os.path.join(cache_db, 'cache-achievements.db')
+        self.data['achievements-by-time'] = dict()
+        self.data['achievements-by-time'] = self.stored_data
+        # self.data = {**self.data, **self.stored_data}
         data = json.dumps(self.data, sort_keys=True,indent=4, separators=(',', ': '))
         with open(achievements_cache_path, 'w') as f:
             f.write(data)
 
+    def check_already_listed(self, buffer_stat, cat):
+        listed = False
+        for i, name in enumerate(d['name'] for d in buffer_stat['children']):
+            if cat == name:
+                listed = True
+        return listed
+
+    def is_end_of_hierarchy(self, j, cat):
+        end = False
+        if j == len(cat)-1:
+                end = True
+        return end
+
+
+    def next_stat_in_hierarchy(self, buffer_stat, cat):
```

Comment:
what if not found? this is a little bit error prone. Please check **all** of you new functions for the error case, including errors you don't think of. E.g. if a if statement is there please add a else statement too and at callee level please handle these errors. If the else case should never happens, please raise an exception - explicit.,